ghcide/src/Development/IDE/Core/WorkerThread.hs

```diff
@@ -42,13 +45,19 @@ withWorkerQueue workerAction = ContT $ \mainAction -> do
                 workerAction l
 
 -- | 'awaitRunInThread' queues up an 'IO' action to be run by a worker thread,
--- and then blocks until the result is computed.
+-- and then blocks until the result is computed. If the action throws an
+-- non-async exception, it is rethrown in the calling thread.
 awaitRunInThread :: TQueue (IO ()) -> IO result -> IO result
 awaitRunInThread q act = do
     -- Take an action from TQueue, run it and
     -- use barrier to wait for the result
     barrier <- newBarrier
     atomically $ writeTQueue q $ do
-        res <- act
-        signalBarrier barrier res
-    waitBarrier barrier
+        resultOrException <- try act
+        case resultOrException of
+            Left e@(fromException -> Just AsyncCancelled) -> throwIO e
+            _ -> signalBarrier barrier resultOrException
```

Comment:
perhaps https://hackage.haskell.org/package/safe-exceptions-0.1.7.4/docs/Control-Exception-Safe.html#v:try is the better `try` here? I don't think we want to catch any kind of `AsyncException` and swallow it.

Reply:
Yes, Nice tip